dawn/docs/parallel_model.rst

```diff
@@ -0,0 +1,251 @@
+
+Formal Definition of the Dawn Parallel Model
+============================================
+
+Contract to the user from HIR
+-----------------------------
+
+Dawn will generate code based on the operations passed from high-level
+intermediate representation. It's specification can be found `here`_.
+
+Dependencies
+~~~~~~~~~~~~
+
+Generally, stencil computation in dawn differentiates between horizontal
+and the vertical dimension. The assumption is that individual statements
+of a stencil are embarrassingly parallel in the horizontal. For that
+reason, there is no guarantee that any horizontal point's computation
+will be finished before any other in each user statement.
+
+Vertical Dependencies
+^^^^^^^^^^^^^^^^^^^^^
+
+Each vertical region of a stencil specifies a loop order of execution.
+Dawn will uphold the following rules:
+
+-  If the loop order is forward, any statement with a vertical
+   dependency on other statements executed on a vertical level with
+   index ``k`` will be executed after each statement that it depends on
+   that appears previously in the current vertical region or in any
+   previous vertical region was executed on every vertical level with
+   indices ``k-1`` and smaller.
+-  If the loop order is backward, any statement with a vertical
+   dependency on another statement executed on a vertical level with
+   index ``k`` will be executed after each statement that it depends on
+   that appears previously in the current vertical region or in any
+   previous vertical region was executed on every vertical level with
+   indices ``k+1`` and bigger.
+
+An example of this would be the following
+
+.. code:: c++
+
+   1 vertical_region(k_start, k_end)
+   2   one = 1
+   3 vertical_region(k_start, k_end)
+   4   two = one[k - 1]
+
+Here, for every level with index ``k``, line two will be executed before
+line 4.
+
+Horizontal Dependencies
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Irrespective of the loop order every statement with a horizontal
+dependency on other statements will be executed after all statements it
+depends on are executed on the full horizontal plane in the same
+vertical level.
```

Comment:
Could we skip the part with "with a horizontal dependency on other statements"? Say that each statement will be executed on the full horizontal plane, but it has the same problem anyway (because of the write-after-read-after-write problem).

Reply:
I don't follow. If you have two pointwise statements
```
a = b
c = a
```
there is no guarantee that `a=b` is executed on the full plane before `c=a` is executed. Of course this guarantee is there on a point by point basis.

Reply:
what I mean is: We can tell the user in the high level representation that each statement is executed on the *full plane* one after the other. Internally, we can still do an analysis of what data he is actually accessing and do it on a point-by-point basis internally, but to the user, it still feels like the full plane is updated one after the other, because the outcome is the same.

Reply:
got it. Can be done